tests/Aspire.Confluent.Kafka.Tests/ProducerConfigurationTests.cs

```diff
@@ -95,14 +95,90 @@ public void ConnectionNameWinsOverConfigSection(bool useKeyed)
+            host.Services.GetRequiredService(ReflectionHelpers.ProducerConnectionFactoryStringKeyStringValueType.Value);
 
-        ProducerConfig config = GetProducerConfig(connectionFactory)!;
+        var config = GetProducerConfig(connectionFactory)!;
 
         Assert.Equal(CommonHelpers.TestingEndpoint, config.BootstrapServers);
     }
 
+    [Theory]
+    [InlineData(true)]
+    [InlineData(false)]
+    public void ConfigureProducerBuilder(bool useKeyed)
+    {
+        var builder = Host.CreateEmptyApplicationBuilder(null);
+        builder.Configuration.AddInMemoryCollection([
+            new KeyValuePair<string, string?>("ConnectionStrings:messaging", CommonHelpers.TestingEndpoint)
+        ]);
+
+        var isCalled = false;
+
+        if (useKeyed)
+        {
+            builder.AddKeyedKafkaProducer<string, string>("messaging", ConfigureBuilder);
+        }
+        else
+        {
+            builder.AddKafkaProducer<string, string>("messaging", ConfigureBuilder);
+        }
+
+        using var host = builder.Build();
+        var connectionFactory = useKeyed ?
+            host.Services.GetRequiredKeyedService(ReflectionHelpers.ProducerConnectionFactoryStringKeyStringValueType.Value, "messaging") :
+            host.Services.GetRequiredService(ReflectionHelpers.ProducerConnectionFactoryStringKeyStringValueType.Value);
+
+        var config = GetProducerConfig(connectionFactory)!;
+
+        Assert.True(isCalled);
+        Assert.Equal(CommonHelpers.TestingEndpoint, config.BootstrapServers);
+        return;
+
+        void ConfigureBuilder(ProducerBuilder<string, string> _)
+        {
+            isCalled = true;
+        }
+    }
+
+    [Theory]
+    [InlineData(true)]
+    [InlineData(false)]
+    public void ConfigureProducerBuilderWithServiceProvider(bool useKeyed)
+    {
+        var builder = Host.CreateEmptyApplicationBuilder(null);
+        builder.Configuration.AddInMemoryCollection([
+            new KeyValuePair<string, string?>("ConnectionStrings:messaging", CommonHelpers.TestingEndpoint)
+        ]);
+
+        var isCalled = false;
+
+        if (useKeyed)
+        {
+            builder.AddKeyedKafkaProducer<string, string>("messaging", ConfigureBuilder);
+        }
+        else
+        {
+            builder.AddKafkaProducer<string, string>("messaging", ConfigureBuilder);
+        }
+
+        using var host = builder.Build();
+        var connectionFactory = useKeyed ?
+            host.Services.GetRequiredKeyedService(ReflectionHelpers.ProducerConnectionFactoryStringKeyStringValueType.Value, "messaging") :
+            host.Services.GetRequiredService(ReflectionHelpers.ProducerConnectionFactoryStringKeyStringValueType.Value);
+
+        var config = GetProducerConfig(connectionFactory)!;
+
+        Assert.True(isCalled);
+        Assert.Equal(CommonHelpers.TestingEndpoint, config.BootstrapServers);
+        return;
+
+        void ConfigureBuilder(IServiceProvider _, ProducerBuilder<string, string> __)
```

Comment:
It would be good if the test actually used the IServiceProvider to ensure it works correctly.

Reply:
done.